### ISSUE TITLE
Internal Class / DateTime object support

### ISSUE THREAD

**Issue by GameplayJDK** [brick/varexporter#10]
Hello there,

I'm currently using `VarExporter` for a caching library.

I'm caching an PHP object graph and just noticed that a DateTime object at any point of the object structure causes an exception to be thrown. Just then I read about the limitation of "internal" classes, which are described to cause that kind of behavior in this library.

So my question is, are there plans to support the DateTime object itself and the "internal" objects related to it (DateInterval and so on)? These classes could be supported by the `SetStateExporter`, as they do implement the `__set_state` method ([php.net](https://www.php.net/manual/en/datetime.set-state.php)).

Currently I'm using a rather ugly approach to avoid this issue: I'm extending the DateTime class with an own non-"internal" class, which for now does the job. But as I said, it's not very pretty to do such stuff. Especially when using the static `createFromFormat()` function, which resulted in following code:

```php
<?php

namespace App;

use App\Workaround\DateTime;
use DateTime as DateTimeInternal;

// ...

/** @var DateTimeInternal $dateTime */
$dateTime = DateTime::createFromFormat(static::DTS_FORMAT, $string) ?: null;

return (new DateTime())
    ->setTimestamp($dateTime->getTimestamp());

// ...
```

```php
<?php

namespace App\Workaround;

use DateTime as DateTimeInternal;

class DateTime extends DateTimeInternal
{
}
```

**EDIT**: Additionally, this workaround does not work, when trying to re-cache an object, as the code produced by `VarExporter` looks as follows:

```php
<?php

// ...

$this->date = \App\Workaround\DateTime::__set_state([
    'date' => '2020-03-03 21:49:31.000000',
    'timezone_type' => 3,
    'timezone' => 'UTC'
]);

// ...
```

This results in yet another DateTime object (the internal one), which cannot be exported afterwards. That forced me to extend all static functions of the internal DateTime object to counter this issue.

**Comment by BenMorel** [comment]
Hi, indeed, `DateTime` and `DateTimeImmutable` can be exported safely by `SetStateExporter`, I just tested that the object can be cast to `(array)` and returns the appropriate values.

What we can do is simply swap `InternalClassExporter` and `SetStateExporter` here:

https://github.com/brick/varexporter/blob/master/src/Internal/GenericExporter.php#L66-L70

I'll happily accept a PR that does this, and adds some tests for `DateTime` and `DateTimeImmutable`.

**Comment by GameplayJDK** [comment]
Hey, thank for the quick reply!

Swapping the export order is better than adding an `instanceof` check inside the `InternalClassExporter`, which was the approach that first came to mind.

I'll do my best to submit that PR as soon as soon as possible.

**Comment by BenMorel** [comment]
Fixed in #11.
